firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java

```diff
@@ -426,15 +429,62 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
               .parseQueryValue(value, op == Operator.IN || op == Operator.NOT_IN);
     }
     FieldFilter filter = FieldFilter.create(fieldPath.getInternalPath(), op, fieldValue);
-    validateNewFilter(filter);
     return filter;
   }
 
+  /**
+   * Takes a {@code CompositeFilterData} object, parses each of its subfilters, and a new {@code
+   * CompositeFilter} that is constructed using the parsed values. Returns null if the given
+   * CompositeFilterData does not contain any subfilters. Returns a {@code FieldFilter} if the given
+   * CompositeFilterData contains only one field filter.
+   */
+  @Nullable
+  private com.google.firebase.firestore.core.Filter parseCompositeFilterData(
+      @NonNull CompositeFilterData compositeFilterData) {
+    List<com.google.firebase.firestore.core.Filter> parsedFilters = new ArrayList<>();
+    for (Filter filter : compositeFilterData.getFilters()) {
+      com.google.firebase.firestore.core.Filter parsedFilter = parseFilterData(filter);
+      if (parsedFilter != null) {
+        parsedFilters.add(parsedFilter);
+      }
+    }
+    // For empty composite filters, return null.
+    // For example: and(or(), and()).
+    if (parsedFilters.isEmpty()) {
+      return null;
+    }
+    // For composite filters containing 1 filter, return the only filter.
+    // For example: AND(FieldFilter1) == FieldFilter1
+    if (parsedFilters.size() == 1) {
+      return parsedFilters.get(0);
+    }
+    return new CompositeFilter(parsedFilters, compositeFilterData.getOperator());
+  }
+
+  /**
+   * Takes a filter whose value has not been parsed, parses the value object and returns a
+   * FieldFilter or CompositeFilter with parsed values.
+   */
+  @Nullable
+  private com.google.firebase.firestore.core.Filter parseFilterData(Filter filter) {
+    hardAssert(
+        filter instanceof FieldFilterData || filter instanceof CompositeFilterData,
+        "Parsing is only supported for FieldFilterData and CompositeFilterData.");
+    if (filter instanceof FieldFilterData) {
+      return parseFieldFilterData((FieldFilterData) filter);
+    }
+    return parseCompositeFilterData((CompositeFilterData) filter);
```

Comment:
I like the idea because it'll get rid of the `instanceof` checking here. But that'd require passing both `FirebaseFirestore` object and the `core/Query` object to the `Filter.parse()` method.

 the signature would have to be the following:
```java
public com.google.firebase.firestore.core.Filter parse(FirebaseFirestore firestore, com.google.firebase.firestore.core.Query query);
```
because the parsing logic needs these.

not sure if it's worth it.